trio/tests/test_exports.py

```diff
@@ -114,6 +143,169 @@ def no_underscores(symbols):
+            continue
+
+        # dir() and inspect.getmembers doesn't display properties from the metaclass
+        # also ignore some dunder methods that tend to differ but are of no consequence
+        ignore_names = set(dir(type(class_))) | {
+            "__annotations__",
+            "__attrs_attrs__",
+            "__attrs_own_setattr__",
+            "__class_getitem__",
+            "__getstate__",
+            "__match_args__",
+            "__order__",
+            "__orig_bases__",
+            "__parameters__",
+            "__setstate__",
+            "__slots__",
+            "__weakref__",
+        }
+
+        # pypy seems to have some additional dunders that differ
+        if sys.implementation.name == "pypy":
+            ignore_names |= {
+                "__basicsize__",
+                "__dictoffset__",
+                "__itemsize__",
+                "__sizeof__",
+                "__weakrefoffset__",
+                "__unicode__",
+            }
+
+        # inspect.getmembers sees `name` and `value` in Enums, otherwise
+        # it behaves the same way as `dir`
+        # runtime_names = no_underscores(dir(class_))
+        runtime_names = (
+            no_hidden(x[0] for x in inspect.getmembers(class_)) - ignore_names
+        )
+
+        if tool == "jedi":
+            import jedi
+
+            script = jedi.Script(
+                f"from {module_name} import {class_name}; {class_name}."
+            )
+            completions = script.complete()
+            static_names = no_hidden(c.name for c in completions) - ignore_names
+
+            missing = runtime_names - static_names
+            extra = static_names - runtime_names
+            if BaseException in class_.__mro__ and sys.version_info > (3, 11):
+                missing.remove("add_note")
+
+            # TODO: why is this? Is it a problem?
+            if class_ == trio.StapledStream:
+                extra.remove("receive_stream")
+                extra.remove("send_stream")
+
+            if missing | extra:
+                errors[f"{module_name}.{class_name}"] = {
+                    "missing": missing,
+                    "extra": extra,
+                }
+        elif tool == "mypy":
+            tmpfile = tmpdir / "check_mypy.py"
+            sorted_runtime_names = list(sorted(runtime_names))
+            content = f"from {module_name} import {class_name}\n" + "".join(
+                f"{class_name}.{name}\n" for name in sorted_runtime_names
+            )
+            tmpfile.write_text(content, encoding="utf8")
+            from mypy.api import run
+
+            res = run(
+                [
+                    "--config-file=",
+                    "--follow-imports=silent",
+                    "--disable-error-code=operator",
+                    str(tmpfile),
+                ]
+            )
+            if res[2] != 0:
+                print(res)
+                it = iter(res[0].split("\n")[:-2])
```

Comment:
There's an option to silence the ending thing. Uhh, here's two flags I think are useful, now that I have the docs up:
 - `--soft-error-limit=-1`
 - `--no-error-summary`

Both can be viewed at https://mypy.readthedocs.io/en/stable/command_line.html#advanced-options

Reply:
oh shit, `--soft-error-limit` is maybe why I had problems when I was running `mypy` and passing the code to check as a command-line parameter instead of printing to a file